### PR TITLE
Additional TLS checks

### PR DESCRIPTION
**Pull request by julek-wolfssl**
- double check which messages need to be encrypted
- check msgs that have to be last in a record

ZD17108

### CODE REVIEW

**Review by JacobBarthelmeh — DISMISSED** [review]

**Comment by JacobBarthelmeh** [comment]
Looks good to me, would like at least one other review on this though since the code changes will affect all TLS/DTLS users.

**Review by dgarske — CHANGES_REQUESTED** [review]
This PR looks good. It grows my build by 384 bytes. Should we consider a way to disable this feature, but keep it on by default?
I'd like Sean to review before its merged.

**Comment by julek-wolfssl** [comment]
I don't think that its a good idea to remove these checks. These are important checks that are relevant for increasing security of all (D)TLS versions.

File: src/internal.c [hunk 1]
```diff
@@ -8950,11 +8951,13 @@ int DtlsMsgSet(DtlsMsg* msg, word32 seq, word16 epoch, const byte* data, byte ty
             WOLFSSL_ERROR_VERBOSE(SEQUENCE_ERROR);
             return SEQUENCE_ERROR;
         }
+        msg->encrypted = msg->encrypted && encrypted ? 1 : 0;
```
Comment (SparkiDev) [inline src/internal.c:8954]: msg->encrypted = msg->encrypted && encrypted;

File: src/internal.c [hunk 2]
```diff
@@ -8950,11 +8951,13 @@ int DtlsMsgSet(DtlsMsg* msg, word32 seq, word16 epoch, const byte* data, byte ty
...
     }
     else {
         msg->type = type;
         msg->epoch = epoch;
         msg->seq = seq;
+        msg->encrypted = encrypted ? 1 : 0;
```
Comment (SparkiDev) [inline src/internal.c:8960]: msg->encrypted = encrypted;

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 1 above)
Comment (julek-wolfssl) [reply src/internal.c]: Fixed

File: src/internal.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/internal.c]: Fixed

**Review by dgarske — CHANGES_REQUESTED** [review]
PR looks great. Can you please add a build option to disable this check for users that might not want it?

**Comment by julek-wolfssl** [comment]
@dgarske build option added with `WOLFSSL_DISABLE_EARLY_SANITY_CHECKS`.

**Review by dgarske — APPROVED** [review]
Over to @SparkiDev to finalize.

**Comment by dgarske** [comment]
Retest this please

**Comment by SparkiDev** [comment]
retest this please

**Review by SparkiDev — APPROVED** [review]

**Review by douzzer — APPROVED** [review]
passed a run through `wolfssl-multi-test.sh ... super-quick-check` (latest sanitizers, `clang-tidy`, and `cppcheck`), rebased on current `master`.

**Comment by dgarske** [comment]
Retest this please

**Pull request merged**